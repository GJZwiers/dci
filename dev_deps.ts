export {
  assert,
  assertEquals,
  assertRejects,
  assertThrows,
} from "https://deno.land/std@0.154.0/testing/asserts.ts";
export {
  assertSnapshot,
} from "https://deno.land/std@0.154.0/testing/snapshot.ts";
