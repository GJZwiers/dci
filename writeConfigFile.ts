import { writeFileSec } from "./writeFileSec.ts";
import { generateJsonc } from "./schema.ts";

export interface Options {
  force: boolean;
  fill: boolean;
  fmt: boolean;
  jsonc: boolean;
  lint: boolean;
  map: boolean;
  name: string;
  task: boolean;
  tsconfig: boolean;
  yes: boolean;
}

export const defaultOpts: Options = {
  force: false,
  fill: false,
  fmt: false,
  jsonc: false,
  lint: false,
  map: false,
  name: "deno.json",
  task: false,
  tsconfig: false,
  yes: false,
};

export type ConfigFile = {
  compilerOptions?: Record<string, unknown>;
  fmt?: {
    files?: {
      include?: string[];
      exclude?: string[];
    };
    options?: Record<string, unknown>;
  };
  lint?: {
    files?: {
      include?: string[];
      exclude?: string[];
    };
    rules?: {
      tags?: string[];
      include?: string[];
      exclude?: string[];
    };
  };
  tasks?: Record<string, unknown>;
  importMap?: string;
};

export async function inputHandler(opts: Options) {
  if (opts.fill || opts.jsonc) {
    opts.name = opts.name.replace(".json", ".jsonc");

    const file = generateJsonc(opts);

    await writeFileSec(
      opts.name,
      new TextEncoder().encode(file),
      {
        force: opts.force,
      },
    );
    return file;
  }

  const configFile: ConfigFile = {};

  if (opts.yes) {
    opts.fmt = true;
    opts.lint = true;
    opts.tsconfig = true;
    opts.task = true;
  }

  if (opts.fmt) {
    configFile.fmt = {
      files: {
        include: [],
        exclude: [],
      },
      options: {},
    };
  }

  if (opts.lint) {
    configFile.lint = {
      files: {
        include: [],
        exclude: [],
      },
      rules: {
        tags: [],
        include: [],
        exclude: [],
      },
    };
  }

  if (opts.map) {
    configFile.importMap = "import_map.json";
  }

  if (opts.task) {
    configFile.tasks = {};
  }

  if (opts.tsconfig) {
    configFile.compilerOptions = {};
  }

  await writeConfigFile(configFile, opts);
  return JSON.stringify(configFile, null, 2);
}

export async function writeConfigFile(
  configFile: ConfigFile,
  opts: Options,
) {
  if (opts.name && !/\.jsonc?$/.test(opts.name)) {
    throw new Error(
      `Chosen filename has an unsupported file extension: ${opts.name}`,
    );
  }

  const denoJson = new TextEncoder()
    .encode(JSON.stringify(configFile, null, 2));

  await writeFileSec(
    `./${opts.name}`,
    denoJson,
    { force: opts.force },
  );
}
