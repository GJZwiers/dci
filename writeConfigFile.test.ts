import { assert, assertEquals, assertSnapshot } from "./dev_deps.ts";
import { defaults, inputHandler, Settings } from "./writeConfigFile.ts";

Deno.test("isSnapshotMatch", async (test) => {
  const config = {
    fmt: {
      files: {
        exclude: [],
        include: []
      },
      options: {},
    },
  };
  await assertSnapshot(test, config);
});

Deno.test("writeConfigFile()", async (context) => {
  const testDir = "test_directory";
  let testSettings: Settings;

  const beforeEach = async () => {
    await Deno.mkdir(testDir, { recursive: true });
    Deno.chdir(testDir);

    testSettings = self.structuredClone(defaults);
  };

  const afterEach = async () => {
    Deno.chdir("..");
    await Deno.remove(testDir, { recursive: true });
    testSettings.name = "deno.json";
    testSettings.map = false;
    testSettings.fill = false;
    testSettings.fmt = false;
  };

  const test = async (
    options: Deno.TestDefinition,
  ) => {
    await beforeEach();

    await context.step(options);

    await afterEach();
  };

  await test({
    name: "create deno.json",
    fn: async () => {
      await inputHandler(testSettings);

      const configFile = await Deno.readFile(
        `${testSettings.name}`,
      );

      assertEquals(testSettings.name, "deno.json");
      assert(configFile);
    },
  });

  await test({
    name: "create filled deno.jsonc if --fill is used",
    fn: async () => {
      testSettings.fill = true;
      await inputHandler(testSettings);

      assertEquals(testSettings.name, "deno.jsonc");

      const bytes = await Deno.readFile(
        `${testSettings.name}`,
      );

      const file = new TextDecoder().decode(bytes);

      assert(file.indexOf("fmt") > 0);
      assert(file.indexOf("lint") > 0);
      assert(file.indexOf("importMap") > 0);
      assert(file.indexOf("compilerOptions") > 0);
      assert(file.indexOf("tasks") > 0);
    },
  });

  await test({
    name:
      "create deno.jsonc with fmt section only if fill is used in combination with fmt",
    fn: async () => {
      testSettings.fill = true;
      testSettings.fmt = true;
      await inputHandler(testSettings);

      const bytes = await Deno.readFile(
        `${testSettings.name}`,
      );

      const file = new TextDecoder().decode(bytes);

      assert(file.indexOf("fmt") > 0);

      assert(file.indexOf("lint") === -1);
      assert(file.indexOf("importMap") === -1);
      assert(file.indexOf("compilerOptions") === -1);
      assert(file.indexOf("tasks") === -1);
    },
  });

  await test({
    name: "create fmt options",
    only: true,
    fn: async () => {
      testSettings.fmt = true;

      await inputHandler(testSettings);

      const configFile = await Deno.readFile(
        `${testSettings.name}`,
      );

      assertEquals(testSettings.name, "deno.json");
      assert(configFile);

      const contents = new TextDecoder().decode(configFile);
      const json = JSON.parse(contents);
      assert(json.fmt);
    },
  });

  await test({
    name: "create lint options",
    fn: async () => {
      testSettings.lint = true;

      await inputHandler(testSettings);

      const configFile = await Deno.readFile(
        `${testSettings.name}`,
      );

      assertEquals(defaults.name, "deno.json");
      assert(configFile);

      const contents = new TextDecoder().decode(configFile);
      const json = JSON.parse(contents);

      assert(json.lint);
    },
  });

  await test({
    name: "create compilerOptions",
    fn: async () => {
      testSettings.tsconfig = true;

      await inputHandler(testSettings);

      const configFile = await Deno.readFile(
        `${testSettings.name}`,
      );

      assertEquals(testSettings.name, "deno.json");
      assert(configFile);

      const contents = new TextDecoder().decode(configFile);
      const json = JSON.parse(contents);

      assert(json.compilerOptions);
    },
  });

  await test({
    name: "create tasks",
    fn: async () => {
      testSettings.task = true;

      await inputHandler(testSettings);

      const configFile = await Deno.readFile(
        `${testSettings.name}`,
      );

      assertEquals(testSettings.name, "deno.json");
      assert(configFile);

      const contents = new TextDecoder().decode(configFile);
      const json = JSON.parse(contents);

      assert(json.tasks);
    },
  });

  await test({
    name: "create importMap",
    fn: async () => {
      testSettings.map = true;

      await inputHandler(testSettings);

      const configFile = await Deno.readFile(
        `${testSettings.name}`,
      );

      assertEquals(testSettings.name, "deno.json");
      assert(configFile);

      const contents = new TextDecoder().decode(configFile);
      const json = JSON.parse(contents);

      assertEquals(json.importMap, "");
    },
  });

  await test({
    name: "create all if yes option is true",
    fn: async () => {
      testSettings.yes = true;

      await inputHandler(testSettings);

      const configFile = await Deno.readFile(
        `${testSettings.name}`,
      );

      assertEquals(testSettings.name, "deno.json");
      assert(configFile);

      const contents = new TextDecoder().decode(configFile);
      const json = JSON.parse(contents);

      assert(json.fmt);
      assert(json.lint);
      assert(json.compilerOptions);
    },
  });
});
