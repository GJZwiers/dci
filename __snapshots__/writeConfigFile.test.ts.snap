export const snapshot = {};

snapshot[`writeConfigFile 1`] = `
[
  "{}",
]
`;

snapshot[`writeConfigFile 2`] = `
[
  "{",
  '  "fmt": {',
  '    "files": {',
  '      "include": [],',
  '      "exclude": []',
  "    },",
  '    "options": {}',
  "  }",
  "}",
]
`;

snapshot[`writeConfigFile 3`] = `
[
  "{",
  '  "compilerOptions": {',
  '    // "allowJs": true                         /* Allow JavaScript files to be a part of your progra...',
  '    // "allowUnreachableCode": false           /* Disable error reporting for unreachable code. */',
  '    // "allowUnusedLabels": false              /* Disable error reporting for unused labels. */',
  '    // "checkJs": false                        /* Enable error reporting in type-checked JavaScript ...',
  '    // "experimentalDecorators": true          /* Enable experimental support for TC39 stage 2 draft...',
  '    // "jsx": "react"                          /* Specify what JSX code is generated. */',
  '    // "jsxFactory": "React.createElement"     /* Specify the JSX factory function used when targeti...',
  '    // "jsxFragmentFactory": "React.Fragment"  /* Specify the JSX Fragment reference used for fragme...',
  '    // "jsxImportSource": "react"              /* Specify module specifier used to import the JSX fa...',
  '    // "keyofStringsOnly": false               /* Make keyof only return strings instead of string, ...',
  '    // "lib": [ "deno.window" ]                /* Specify a set of bundled library declaration files...',
  '    // "noFallthroughCasesInSwitch": false     /* Enable error reporting for fallthrough cases in sw...',
  '    // "noImplicitAny": true                   /* Enable error reporting for expressions and declara...',
  '    // "noImplicitOverride": false             /* Ensure overriding members in derived classes are m...',
  '    // "noImplicitReturns": false              /* Enable error reporting for codepaths that do not e...',
  '    // "noImplicitThis": true                  /* Enable error reporting when \`this\` is given the ty...',
  \`    // "noImplicitUseStrict": true             /* Disable adding 'use strict' directives in emitted ...\`,
  '    // "noStrictGenericChecks": false          /* Disable strict checking of generic signatures in f...',
  '    // "noUnusedLocals": false                 /* Enable error reporting when a local variables aren...',
  \`    // "noUnusedParameters": false             /* Raise an error when a function parameter isn't rea...\`,
  '    // "noUncheckedIndexedAccess": false       /* Add \`undefined\` to a type when accessed using an i...',
  '    // "strict": true                          /* Enable all strict type checking options. */',
  '    // "strictBindCallApply": true             /* Check that the arguments for \`bind\`, \`call\`, and \`...',
  '    // "strictFunctionTypes": true             /* When assigning functions, check to ensure paramete...',
  '    // "strictPropertyInitialization": true    /* Check for class properties that are declared but n...',
  '    // "strictNullChecks": true                /* When type checking, take into account \`null\` and \`...',
  '    // "suppressExcessPropertyErrors": false   /* Disable reporting of excess property errors during...',
  '    // "suppressImplicitAnyIndexErrors": false /* Suppress \`noImplicitAny\` errors when indexing obje...',
  "  },",
  '  // "importMap": "",                          /* The location of an import map to be used when reso...',
  '  "lint": {',
  '    "files": {',
  '      // "include": []                         /* List of files or directories that will be linted. ...',
  '      // "exclude": []                         /* List of files or directories that will not be lint...',
  "    },",
  '    "rules": {',
  '      // "tags": []                            /* List of tag names that will be run. Empty list dis...',
  '      // "exclude": []                         /* List of rule names that will be excluded from conf...',
  '      // "include": []                         /* List of rule names that will be run. Even if the s...',
  "    }",
  "  },",
  '  "fmt": {',
  '    "files": {',
  '      // "include": []                         /* List of files or directories that will be formatte...',
  '      // "exclude": []                         /* List of files or directories that will not be form...',
  "    },",
  '    "options": {',
  '      // "useTabs": false                      /* Whether to use tabs (true) or spaces (false) for i...',
  '      // "lineWidth": 80                       /* The width of a line the printer will try to stay u...',
  '      // "indentWidth": 2                      /* The number of characters for an indent. */',
  '      // "singleQuote": false                  /* Whether to use single quote (true) or double quote...',
  '      // "proseWrap": "always"                 /* Define how prose should be wrapped in Markdown fil...',
  "    }",
  "  },",
  '  "tasks": {}',
  "}",
]
`;

snapshot[`writeConfigFile 4`] = `
[
  "{",
  '  "fmt": {',
  '    "files": {',
  '      // "include": []         /* List of files or directories that will be formatted. */',
  '      // "exclude": []         /* List of files or directories that will not be formatted. */',
  "    },",
  '    "options": {',
  '      // "useTabs": false      /* Whether to use tabs (true) or spaces (false) for indentation. */',
  '      // "lineWidth": 80       /* The width of a line the printer will try to stay under. Note that ...',
  '      // "indentWidth": 2      /* The number of characters for an indent. */',
  '      // "singleQuote": false  /* Whether to use single quote (true) or double quote (false) for quo...',
  '      // "proseWrap": "always" /* Define how prose should be wrapped in Markdown files. */',
  "    }",
  "  }",
  "}",
]
`;

snapshot[`writeConfigFile 5`] = `
[
  "{",
  '  "lint": {',
  '    "files": {',
  '      "include": [],',
  '      "exclude": []',
  "    },",
  '    "rules": {',
  '      "tags": [],',
  '      "include": [],',
  '      "exclude": []',
  "    }",
  "  }",
  "}",
]
`;

snapshot[`writeConfigFile 6`] = `
[
  "{",
  '  "compilerOptions": {}',
  "}",
]
`;

snapshot[`writeConfigFile 7`] = `
[
  "{",
  '  "tasks": {}',
  "}",
]
`;

snapshot[`writeConfigFile 8`] = `
[
  "{",
  '  "importMap": "import_map.json"',
  "}",
]
`;

snapshot[`writeConfigFile 9`] = `
[
  "{",
  '  "fmt": {',
  '    "files": {',
  '      "include": [],',
  '      "exclude": []',
  "    },",
  '    "options": {}',
  "  },",
  '  "lint": {',
  '    "files": {',
  '      "include": [],',
  '      "exclude": []',
  "    },",
  '    "rules": {',
  '      "tags": [],',
  '      "include": [],',
  '      "exclude": []',
  "    }",
  "  },",
  '  "tasks": {},',
  '  "compilerOptions": {}',
  "}",
]
`;
